feat: Add course selection modal to catalog page

aarowz opened the pull request:
Summary
     - Implements the Add Courses modal on the catalog page with full course selection functionality.
     
Changes
     - New `AddCoursesModal` component with two-column layout
     - Real-time course search via `/api/search` endpoint
     - Course selection (add/remove, max 10 courses)
     - Lock/unlock toggle for selected courses
     - Generate schedules button (navigates to scheduler page)
     - Term display name conversion (shows "Spring 2026" instead of "202630")
     - Loading states and error handling
     - Backend API integration with debounced search
     
Testing
     - [x] Modal opens from catalog page
     - [x] Course search works with real API data
     - [x] Add/remove courses functionality
     - [x] Lock/unlock toggle works
     - [x] Generate schedules navigates correctly
     - [x] Term display name shows correctly
     
Future Work (not in this PR)
     - Campus filtering (placeholder implemented)
     - Child sections (recitations/labs)
     - Info button functionality

![0FCC8BE5-4157-4198-84D9-8260A10C710B](https://github.com/user-attachments/assets/ca378109-6708-469d-b6f9-1a2155dadd81)

pranavphadke1 commented:
Note: 
- Will probably use this error page as default. Should prob change the page name to something cleaner than "/apiError"
- I believe the only "error" we can get from the useSearch function in "pages/campus/termid/search/query.tsx" is "TypeError: Network request failed", which I believe can only happen if the API isn't working. If this is the case, maybe we should just redirect to this new error page in a similar fasion as we do in the homepage (termid.tsx): 
- useEffect(router.push(apiErrorPage), [error]);

pranavphadke1 commented:
Need to do:
- Rename page to "error"
- Implement Vanessa's new Button design
- Experiment with z-index to fix mobile page viewing size issue
- Redirect to error page when searching with API down
- Console log the specific error we run into

pranavphadke1 commented:
These are the console log messages I added @hankewyczz 

When redirected to error page from the main page:
<img width="323" alt="image" src="https://user-images.githubusercontent.com/91645231/218337781-1b964ccf-b9a4-45b4-a5b7-4c9509bfd107.png">

When redirected to error page from a failed search:
<img width="321" alt="image" src="https://user-images.githubusercontent.com/91645231/218337858-4e7fa597-c32c-4920-a483-bfcf0d7185a0.png">

hankewyczz commented on pages/error.tsx:
```diff
@@ -0,0 +1,84 @@
+
+  const [modalOpen, setModalOpen] = useState(false);
+
+  const toggleModal = () => {
+    setModalOpen(!modalOpen);
+  };
```
```suggestion
  const toggleModal = () => setModalOpen(!modalOpen);
```
a small nit

hankewyczz commented on pages/error.tsx:
```diff
@@ -0,0 +1,84 @@
+  };
+
+  return (
+    <>
```
Can we not just use the top-level `div` here?

hankewyczz commented on pages/error.tsx:
```diff
@@ -0,0 +1,84 @@
+                    target="_blank"
+                    rel="noopener noreferrer"
+                    className="create-issue"
+                    href="https://github.com/sandboxnu/searchneu"
```
```suggestion
                    href="https://github.com/sandboxnu/searchneu/issues/new"
```

hankewyczz commented on styles/pages/_Error.scss:
```diff
@@ -0,0 +1,158 @@
+  bottom: -40px;
+  z-index: 102;
+
+  @media only screen and (min-width: 0px) {
+    // Screen too small to display
+    width: 0%;
+  }
+
+  @media only screen and (min-width: 200px) {
+    width: 45%;
+    right: 0px;
+    bottom: -10px;
+  }
```
I'm not a fan of this pattern — I think it could be simplified. 

I think the ideal setup would be using flex columns — https://css-tricks.com/snippets/css/a-guide-to-flexbox/

When the screen is wide enough, it's a grid with two columns, one row (the husky in the rightmost cell)

When it isn't, it's a grid with one column and two rows (husky in the bottom cell).

Tbh, we could use flexbox for more of the elements here, like arranging the buttons. This shouldn't be something we have to create on our own.

pranavphadke1 commented:
@hankewyczz Redid the CSS using flexbox. This helped remove a lot of hard-coded styling and also got rid of the need for z-index.

Overall, reduced majority of @ media calls. However, they are still some necessary usage such as adding flex-wrap to the "error-container" class. Also needed to use it to handle the vertical orientation on mobile screens.

pranavphadke1 commented:
@Lucas-Dunker ready for review

Lucas-Dunker reviewed (approved):
Looks good to me!

sebwittr commented on pages/[campus]/[termId].tsx:
```diff
@@ -39,11 +38,12 @@ export default function Home(): ReactElement {
-  };
+  useEffect(() => {
+    if (termInfosError) {
+      console.log("We've encountered an error: " + termInfosError);
```
You should get rid of any console.logs you accidentally leave in.

pranavphadke1 replied:
These are to let users know what specific error occured (if they're curious). This was something Zachar asked to add. Is this how you want this implemented? @hankewyczz ?

sebwittr replied:
ok cool sounds good then!

sebwittr commented on pages/error.tsx:
```diff
@@ -0,0 +1,87 @@
+                  <div className="request-container">
+                    <button
+                      type="button"
+                      className="suport-button"
```
typo: probably should be "support button"

sebwittr reviewed (approved):
Code looks good! I noticed a typo in the code, but it doesn't matter too much.

hankewyczz commented on styles/pages/_Error.scss:
```diff
@@ -0,0 +1,150 @@
+.error-container {
+  height: 100vh;
+  width: 100vw;
+  display: flex !important;
```
Try to make sure that all `!important` uses are actually needed — if not, we don't want to introduce too many (not saying this one is unnecessary, just asking for a second lookthrough)

hankewyczz commented on styles/pages/_Error.scss:
```diff
@@ -0,0 +1,150 @@
+}
+
+.api-error-title-text {
+  font-family: Montserrat Alternates;
```
```suggestion
  font-family: 'Montserrat Alternates';
```

hankewyczz reviewed (approved):
Approved, but two quick things first:
- Try to remove any `!important` cases which aren't necessary
- Double-check with Vanessa (send her the staging link) that the design is OK

Looks great!

aarowz committed:
feat: Add course selection modal to catalog page

- Implement AddCoursesModal component with real-time course search
- Integrate modal into catalog page with 'Add Courses' button
- Add backend API integration for course fetching
- Implement course selection (add/remove, max 10 courses)
- Add lock/unlock functionality for selected courses
- Convert term ID to display name (e.g., 'Spring 2026')
- Add loading states and error handling
- Enable scheduler flag for navigation to scheduler page

mehallhm reviewed (changes requested):
Couple more remarks:
- not sure what the Figma says but the selected color for a course in the search results list is too light
- if the course name is long, then the icons akwardly cover the end of the course title. prob should just truncate the length
- weird behavior with long course registers and titles. also think you should properly truncate, but discuss w/ design
- search icon in the search bar is facing the wrong way
- the modal does not have the default animations in or out?
- not sure if hotkeys are meant to work to exit (ie `esc`, clicking off the modal, etc.)
- state persists in the modal if it is open then closed
- bottom of the search result list and right pane do not line up
- accessibility is a nightmare... the icons for more info and to delete the course do not show up when tabbing over them
- focus is not stollen from the background content (can still tab through it)
- ironically can still press `/` to search (need to clear the event handler)

I am not really involved as much as I would like to be for plan but is that the best spot of the "Add Courses" button? Feels... off

kaiyangzhengklaviyo committed:
address comments

kaiyangzhengklaviyo committed:
fix

kaiyangzhengklaviyo committed:
fix

kaiyangzhengklaviyo committed:
merge
